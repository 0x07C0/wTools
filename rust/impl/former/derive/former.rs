// #![allow( unused_imports )]
#![ warn( missing_docs ) ]

use quote::quote;
use syn::{ DeriveInput };
// use proc_macro_error::{ abort };
use unzip_n::unzip_n;

use crate::wproc_macro;
// #[macro_use]
// extern crate wproc_macro;

unzip_n!( 0 );
unzip_n!( 3 );
unzip_n!( 4 );
unzip_n!( 5 );

/* xxx :
- remove unwraps
- add documentation generation
- add documentation
*/

//

#[allow( dead_code )]
struct FormerField< 'a >
{
  pub attrs : &'a Vec< syn::Attribute >,
  pub vis : &'a syn::Visibility,
  pub ident : &'a Option< syn::Ident >,
  pub colon_token : &'a Option< syn::token::Colon >,
  pub ty : &'a syn::Type,
  pub non_optional_ty : &'a syn::Type,
  pub is_option : bool,
  pub type_container_kind : wproc_macro::ContainerKind,
}

//

fn is_option( ty : &syn::Type ) -> bool
{
  wproc_macro::type_rightmost( ty ) == Some( "Option".to_string() )
}

//

fn parameter_internal_first( ty : &syn::Type ) -> &syn::Type
{
  wproc_macro::type_parameters( ty, 0 ..= 0 )
  .first()
  .or_else( || panic!( "Expect at least one parameter here:\n  {}", quote!{ #ty } ) ).unwrap()
}

//

fn parameter_internal_first_two( ty : &syn::Type ) -> Result< ( &syn::Type, &syn::Type ), syn::Error >
{
  let on_err = ||
  {
    syn::Error::new( proc_macro2::Span::call_site(), format!( "Expect at least two parameters here:\n  {}", quote!{ #ty } ) )
  };
  let result = wproc_macro::type_parameters( ty, 0 ..= 1 );
  let mut iter = result.iter();
  Ok
  ((
    iter.next().ok_or_else( on_err )?,
    iter.next().ok_or_else( on_err )?,
  ),)
}

//

#[inline]
fn field_none_map( field : &FormerField ) -> syn::Field
{
  let ident = field.ident.clone();
  let colon_token = field.colon_token.clone();
  let tokens = quote! { ::core::option::Option::None };
  let ty2 : syn::Type = syn::parse2( tokens ).unwrap();
  syn::Field
  {
    attrs : vec![],
    vis : syn::Visibility::Inherited,
    ident,
    colon_token,
    ty : ty2,
  }
}

//

#[inline]
fn field_optional_map( field : &FormerField ) -> syn::Field
{
  let ident = field.ident.clone();
  let colon_token = field.colon_token.clone();
  let ty = field.ty.clone();

  let tokens = if is_option( &ty )
  {
    quote! { #ty }
  }
  else
  {
    quote! { ::core::option::Option< #ty > }
  };

  let ty2 : syn::Type = syn::parse2( tokens ).unwrap();

  syn::Field
  {
    attrs : vec![],
    vis : syn::Visibility::Inherited,
    ident,
    colon_token,
    ty : ty2,
  }
}

//

#[inline]
fn field_form_map( field : &FormerField ) -> syn::Stmt
{
  let ident = field.ident;
  let ty = field.ty;

  let tokens = if field.is_option
  {
    quote!
    {
      let #ident = if self.#ident.is_some()
      {
        ::core::option::Option::Some( self.#ident.take().unwrap() )
      }
      else
      {
        ::core::option::Option::None
      };
    }
  }
  else
  {
    quote!
    {
      let #ident = if self.#ident.is_some()
      {
        self.#ident.take().unwrap()
      }
      else
      {
        let val : #ty = ::core::default::Default::default();
        val
      };
    }
  };

  let stmt : syn::Stmt = syn::parse2( tokens ).unwrap();
  stmt
}

//

#[inline]
fn field_name_map( field : &FormerField ) -> syn::Ident
{
  let ident = field.ident.clone();
  if let Some( ident ) = ident
  {
    ident
  }
  else
  {
    syn::Ident::new( "?? no name_ident ??", proc_macro2::Span::call_site() )
  }
}

//

#[inline]
fn field_setter_map( field : &FormerField, former_name : &syn::Ident ) -> Result< syn::Stmt, syn::Error >
{
  let ident = field.ident.clone();

  let tokens = match &field.type_container_kind
  {
    wproc_macro::ContainerKind::No =>
    {
      let non_optional_ty = &field.non_optional_ty;
      quote!
      {
        #[inline]
        pub fn #ident< Src >( mut self, src : Src ) -> Self
        where Src : ::core::convert::Into< #non_optional_ty >,
        {
          debug_assert!( self.#ident.is_none() );
          self.#ident = ::core::option::Option::Some( src.into() );
          self
        }
      }
    },
    wproc_macro::ContainerKind::Vector =>
    {
      let ty = &field.ty;
      let internal_ty = parameter_internal_first( ty );
      quote!
      {
        #[inline]
        pub fn #ident( mut self ) -> former::runtime::VectorFormer
        <
          #internal_ty,
          #ty,
          #former_name,
          impl Fn( &mut #former_name, ::core::option::Option< #ty > )
        >
        {
          let container = self.#ident.take();
          let on_end = | former : &mut #former_name, container : ::core::option::Option< #ty > |
          {
            former.#ident = container;
          };
          former::runtime::VectorFormer::new( self, container, on_end )
        }
      }
    },
    wproc_macro::ContainerKind::HashMap =>
    {
      let ty = &field.ty;
      let ( k_ty, e_ty ) = parameter_internal_first_two( ty )?;
      quote!
      {
        #[inline]
        pub fn #ident( mut self ) -> former::runtime::HashmapFormer
        <
          #k_ty,
          #e_ty,
          #ty,
          #former_name,
          impl Fn( &mut #former_name, ::core::option::Option< #ty > )
        >
        {
          let container = self.hashmap_strings_1.take();
          let on_end = | former : &mut #former_name, container : ::core::option::Option< #ty > |
          {
            former.hashmap_strings_1 = container;
          };
          former::runtime::HashmapFormer::new( self, container, on_end )
        }
      }
    },
  };

  let stmt : syn::Stmt = syn::parse2( tokens ).unwrap();
  Ok( stmt )

  // pub fn int_1< Src >( mut self, src : Src ) -> Self
  // where Src : Into< i32 >,
  // {
  //   debug_assert!( self.int_1.is_none() );
  //   self.int_1 = Some( src.into() );
  //   self
  // }

}

//

// use syn::spanned::Spanned;
// use wproc_macro::DataWrapped;

pub fn former( input : proc_macro::TokenStream ) -> Result< proc_macro2::TokenStream, syn::Error >
{
  // let ast = parse_macro_input!( input as DeriveInput );

  let ast = match syn::parse::< DeriveInput >( input )
  {
    Ok( syntax_tree ) => syntax_tree,
    Err( err ) => return Err( err ),
  };

  let name_ident = &ast.ident;
  let former_name = format!( "{}Former", name_ident );
  let former_ident = syn::Ident::new( &former_name, name_ident.span() );
  let fields = if let syn::Data::Struct( syn::DataStruct { fields : syn::Fields::Named( syn::FieldsNamed { ref named, .. } ), .. } ) = ast.data
  {
    named
  }
  else
  {
    let data = &ast.data;
    return Err( syn::Error::new( wproc_macro::span_of( data ), format!( "Expects struct, found :\n{}", "xxx"/*quote!{ #data } */ ) ) );
    // return Err( syn::Error::new( wproc_macro::DataWrapped( data ).span(), format!( "Expects struct, found :\n{}", "xxx"/*quote!{ #data } */ ) ) );
    // return Err( syn::Error::new( data.span(), format!( "Expects struct, found :\n{}", "xxx"/*quote!{ #data } */ ) ) );
    // return Err( syn::Error::new( proc_macro2::Span::call_site(), format!( "Expects struct, found :\n{}", "xxx"/*quote!{ #data } */ ) ) );
  };

  let( fields_none, fields_optional, fields_form, fields_names, fields_setter ) = fields.iter().map( | field |
  {
    let attrs = &field.attrs;
    let vis = &field.vis;
    let ident = &field.ident;
    let colon_token = &field.colon_token;
    let ty = &field.ty;
    let is_option = is_option( &ty );
    let type_container_kind = wproc_macro::type_container_kind( &ty );
    let non_optional_ty : &syn::Type = if is_option { parameter_internal_first( ty ) } else { ty };
    let former_field = FormerField { attrs, vis, ident, colon_token, ty, non_optional_ty, is_option, type_container_kind };
    (
      field_none_map( &former_field ),
      field_optional_map( &former_field ),
      field_form_map( &former_field ),
      field_name_map( &former_field ),
      field_setter_map( &former_field, &former_ident ),
    )
  }).unzip_n_vec();

  if let Some( Err( err ) ) = fields_setter.iter().find( | e | e.is_err() )
  {
    return Err( err.clone() );
  }

  let fields_setter = fields_setter.into_iter().map( | e | e.unwrap() ).collect::< Vec< syn::Stmt > >();

  let result = quote!
  {

    impl #name_ident
    {
      #[inline]
      pub fn former() -> #former_ident
      {
        #former_ident
        {
          #( #fields_none, )*
        }
      }
    }

    /* qqq : pub is optional here */
    #[derive( Debug )]
    pub struct #former_ident
    {
      #( #fields_optional, )*
    }

    impl #former_ident
    {
      #[inline]
      pub fn form( mut self ) -> #name_ident
      {
        #( #fields_form )*
        #name_ident
        {
          #( #fields_names, )*
        }
      }
      #( #fields_setter )*
    }

  };

  // println!( "{:#?}", ast );
  // println!( "{:#?}", result );
  // let result = proc_macro2::TokenStream::new();
  // result.into()
  Ok( result )
}
