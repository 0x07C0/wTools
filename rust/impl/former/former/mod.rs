#![ warn( missing_docs ) ]

//!
//! Former - variation of builder pattern.
//!
//! # Sample
//! ```
//! use former::Former;
//!
//! #[derive( Debug, PartialEq, Former )]
//! pub struct Command
//! {
//!   int_1 : i32,
//!   string_1 : String,
//!   vec_1 : Vec< String >,
//!   hashmap_strings_1 : std::collections::HashMap< String, String >,
//!   int_optional_1 : core::option::Option< i32 >,
//!   string_optional_1 : Option< String >,
//! }
//!
//! fn main()
//! {
//!
//!   let command = Command::former()
//!   .int_1( 13 )
//!   .string_1( "Abcd".to_string() )
//!   .vec_1().push( "ghi" ).push( "klm" ).end()
//!   .hashmap_strings_1().insert( "k1", "v1" ).insert( "k2", "v2" ).end()
//!   .string_optional_1( "dir1" )
//!   .form();
//!   dbg!( &command );
//!
//! }
//! ```

// #![ feature( type_name_of_val ) ]
// #![ feature( trace_macros ) ]

// pub extern crate former_derive;
// pub extern crate former_runtime;

pub use former_derive as derive;
pub use former_runtime as runtime;
pub use derive::Former;

// #[macro_use]
// pub extern crate maplit;
