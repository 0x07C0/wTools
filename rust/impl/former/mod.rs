// #![ feature( type_name_of_val ) ]
// #![ feature( trace_macros ) ]

// pub extern crate former_derive;
// pub extern crate former_runtime;

// #[macro_use]
// pub extern crate maplit;

pub mod runtime;
// pub mod derive;
// pub mod former;

// pub use runtime as former_runtime;
// pub use derive as former_derive;
