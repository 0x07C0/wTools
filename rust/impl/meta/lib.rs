#![ warn( missing_docs ) ]
#![ warn( missing_debug_implementations ) ]

//!
//! Collection of general purpose meta tools.
//!

///
/// Collection of general purpose meta tools.
///

pub mod meta
{
  // pub use maplit::*;
  pub use literally::*;
}

pub use meta::*;
