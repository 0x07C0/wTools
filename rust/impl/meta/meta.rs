/// Internal namespace.
pub( crate ) mod private
{

}

//

// qqq : xxx : optimize
mod_interface::mod_interface!
{

  #[ cfg( feature = "impls_index" ) ]
  prelude use ::impls_index::prelude::*;
  #[ cfg( feature = "mod_interface" ) ]
  prelude use ::mod_interface::prelude::*;
  prelude use ::mod_interface::prelude::mod_interface;

  #[ cfg( feature = "for_each" ) ]
  prelude use ::for_each::prelude::*;
  #[ cfg( any( feature = "options", feature = "meta_options" ) ) ]
  prelude use ::woptions::prelude::*;
  #[ cfg( any( feature = "options", feature = "meta_options" ) ) ]
  prelude use ::woptions as options;
  #[ cfg( any( feature = "former", feature = "meta_former" ) ) ]
  prelude use ::former::prelude::*;
  #[ cfg( any( feature = "former", feature = "meta_former" ) ) ]
  prelude use ::former as former;
  #[ cfg( feature = "collection_make" ) ]
  prelude use ::literally::*;

}
