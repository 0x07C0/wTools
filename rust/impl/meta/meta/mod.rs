/// Internal namespace.
pub mod internal
{

}

/// Dependencies.
pub mod dependencies
{
  pub use ::literally;
  pub use ::either;
}

/// Several macro to encourage write readable code.
pub mod impls;

/* xxx2 : use name protected */
/* zzz : use for implementing of macro mod_interface */

/// Owned namespace of the module.
pub mod own
{
  pub use super::parented::*;
}

pub use own::*;

/// Shared with parent namespace of the module
pub mod parented
{
  pub use super::exposed::*;
  pub use super::dependencies;
}

/// Exposed namespace of the module.
pub mod exposed
{
  pub use super::prelude::*;
  pub use super::impls::exposed::*;
}

/// Prelude to use: `use wtools::prelude::*`.
pub mod prelude
{
  pub use literally::*;
  pub use either::*;
  pub use super::impls::prelude::*;
}
