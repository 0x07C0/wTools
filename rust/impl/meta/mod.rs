
pub use is_slice::*;
pub use instance_of::*;
pub use inspect_type::*;