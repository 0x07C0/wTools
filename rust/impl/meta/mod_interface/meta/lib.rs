#![ warn( rust_2018_idioms ) ]
#![ warn( missing_debug_implementations ) ]
#![ warn( missing_docs ) ]

// #![ feature( type_name_of_val ) ]
// #![ feature( trace_macros ) ]

//!
//! Protocol of modularity unifying interface of a module.
//!

#![ doc = include_str!( concat!( env!( "CARGO_MANIFEST_DIR" ), "/Readme.md" ) ) ]

mod impls;

///
/// Protocol of modularity unifying interface of a module.
///

#[ proc_macro ]
pub fn mod_interface( input : proc_macro::TokenStream ) -> proc_macro::TokenStream
{
  let result = impls::impls( input );
  match result
  {
    Ok( stream ) => stream.into(),
    Err( err ) => err.to_compile_error().into(),
  }
}

/*

mod_interface!
{

  pub mod file1;
  pub mod file2;

  private mod micro_private;
  protected mod micro_protected;
  orphan mod micro_orphan;
  exposed mod micro_exposed;
  prelude mod micro_prelude;

  use prelude_file::*;

}

      private      <      protected      <      orphan      <      exposed      <      prelude
      itself               itself             its parent       its inter-module    its inter-module
      private              public               public             public              public

micro-module < meso-module < macro-module < inter-module

slim module - ?
complete module - ?

*/

