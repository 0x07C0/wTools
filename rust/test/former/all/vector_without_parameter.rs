#[ allow( unused_imports ) ]
use super::*;

only_for_wtools!
{
  #[ allow( unused_imports ) ]
  use wtools::meta::*;
  #[ allow( unused_imports ) ]
  use wtools::former::Former;
}

only_for_local_module!
{
  #[ allow( unused_imports ) ]
  use meta_tools::*;
  #[ allow( unused_imports ) ]
  use former::Former;
}

struct Vec
{
}

#[derive( Former )]
pub struct Struct1
{
  pub string_slice_1 : Vec<>,
}

fn main()
{
}
