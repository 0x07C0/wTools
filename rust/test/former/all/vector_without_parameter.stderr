error: Expects at least one parameter here:
         Vec < >
  --> ../../../rust/test/former/all/vector_without_parameter.rs:14:24
   |
14 |   pub string_slice_1 : Vec<>,
   |                        ^^^^^
